### PR TITLE
Include audit logging

### LINKED ISSUE #218: Research django-audit and how audits of models like Application work
_Outcome_: Select a process for auditing events in our Django application, investigate django-audit 

- [ ] Write up ADR

# Background

It is a good idea to have traditional domain models like “Application” and some way to tie LoginUser to Application. Whereas the registry will be complete source of truth for Domains, we will have to track and keep the truth for Applications.

There are two areas event sourcing or even tracking could bring benefit - Application (just a series of events) and changes to users permissions/roles/domains/etc. We believe this will be best modeled in the db with traditional modeling for it and decouple the audit logs - a separate table that is an event source but it not the model.

The benefits to this approach: 
- For analysts working on an application, see a history of how and when and who has changed the state of that application 
- The detail page for a domain application would have a table for past changes - SELECT FROM events WHERE app.id == app.id

### PR DESCRIPTION
**Pull request by SSPJ**
## 🗣 Description ##

The purpose of this PR is to automatically log any model changes to an audit table in the database.

It also includes a refactoring of models.py. To make this work easier to review, changes to the models are in one commit and then moving the models to their new file locations are in a subsequent commit. (Except for a couple of tiny bug fixes while I was getting the tests to run.)

@neilmb I'm tagging you as primary reviewer, but @igorkorenfeld this also touches on your work and I'd welcome any feedback.

## 💭 Motivation and context ##

As implemented, this logs _all_ changes, which includes keeping an audit log of the `last_login` field of the User table. This may not be desired.

Closes #218

## How to test ##

This requires a full dev reset. The suggested way to do this:

```
docker-compose down
docker-compose build --no-cache
docker-compose up
(in other window) docker-compose exec app ./manage.py loaddata registrar/fixtures/*
```

Go to http://localhost:8080/edit_profile and edit the display name. Go to http://localhost:8080/admin/auditlog/logentry/ and verify that the change was recorded.

Go to http://localhost:8080/admin/registrar/domainapplication/add/ and add a domain application. Go to the detail view for the newly added application and click "History". Verify that you see the changes you've made.

image

image

### CODE REVIEW

File: src/registrar/models/domain_application.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+        blank=True,
+        help_text="Acknowledged .gov acceptable use policy",
+    )
+
+    def __str__(self):
+        try:
+            if self.requested_domain and self.requested_domain.website:
+                return self.requested_domain.website
+            else:
+                return f"{self.status} application created by {self.creator}"
```
Comment (neilmb) [inline src/registrar/models/domain_application.py:210]: This is a lovely default for when the site is not set.

File: src/registrar/admin.py [hunk 2]
```diff
@@ -18,3 +37,5 @@ class MyUserAdmin(UserAdmin):
 
 
 admin.site.register(User, MyUserAdmin)
+admin.site.register(DomainApplication, AuditedAdmin)
```
Comment (neilmb) [inline src/registrar/admin.py:40]: **question (non-blocking):** Do we need/want tests on our custom admin views? Is it easy enough to make a `TestCase` that makes an admin user, uses `force_login`, and loads one or more of these admin pages?

**Review by neilmb — APPROVED** [review]
LGTM, great work on the `models` module and a truly non-blocking question about testing admin views.

File: src/registrar/admin.py (hunk 2 above)
Comment (SSPJ) [reply src/registrar/admin.py:40]: I think that's a great idea, but I'm just going to make a note on #14 about it for the moment.

**Pull request merged**